### DISCUSSION TITLE
What C++ version(s) should our C++ backend require?

### DISCUSSION THREAD

**Discussion by steven-johnson** [halide/Halide#5311] [General]
It looks like our C++ backend generates code that tries to assume only C++98 (e.g. it inserts `static_assert` calls bracketed with checks of `__cplusplus >= 201103L`).

Should we consider upgrading the requirement for the code we *generate*, to be at least C++11? 

(Note: we should really stop calling it the "C backend" as it hasn't been plain C in, uh, ever?)

**Comment by vksnk** [comment]
I think it's already using C++11 in a few places, for example:

- nullptr (https://github.com/halide/Halide/blob/master/src/CodeGen_C.cpp#L209)
- constexpr (https://github.com/halide/Halide/blob/master/src/CodeGen_C.cpp#L171)

At some point I tried to compile generated code with C++98 and it didn't work right away (mostly because of what mentioned above), but it wasn't too difficult to fix (and I sort of already assumed that C++11 is assumed, so just fixed it locally).
